### PR TITLE
III-5334 messaging contributors

### PR DESCRIPTION
**Pull request by JonasVHG**
### Added

- `BroadcastingContributorRepository` as `Decorator` to publish messages
- Abstract `ContributorsUpdated` And childClasses `EventContributorsUpdated`, `PlaceContributorsUpdated` & `OrganizerContributorsUpdated` for Messages
- `ContributorsUpdatedFactory` to create message depending on `ItemType`

### Changed

- Use `BroadcastingContributorRepository` instead of `DbalContributorRepository` in `ServiceProvider`

---
Ticket: https://jira.uitdatabank.be/browse/III-5334

### CODE REVIEW

File: src/Contributor/BroadcastingContributorRepository.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    public function isContributor(UUID $id, EmailAddress $emailAddress): bool
+    {
+        return $this->repository->isContributor($id, $emailAddress);
+    }
+
+    public function updateContributors(UUID $id, EmailAddresses $emailAddresses, ItemType $itemType): void
+    {
+        $this->repository->updateContributors($id, $emailAddresses, $itemType);
+
+        if ($itemType->sameAs(ItemType::event())) {
```
Comment (LucWollants) [inline src/Contributor/BroadcastingContributorRepository.php:47]: Deciding which event needs to be created should be done inside the factory based on the `ItemType`. So the `ContributorsUpdatedFactory` should have one method `createForItemType(ItemType $itemType)` which contains early returns for every event type.

File: src/Contributor/ContributorsUpdatedFactory.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+namespace CultuurNet\UDB3\Contributor;
+
+use CultuurNet\UDB3\Event\EventContributorsUpdated;
+use CultuurNet\UDB3\Iri\IriGeneratorInterface;
+use CultuurNet\UDB3\Organizer\OrganizerContributorsUpdated;
+use CultuurNet\UDB3\Place\PlaceContributorsUpdated;
+
+final class ContributorsUpdatedFactory
+{
+    private IriGeneratorInterface $eventContributorsUpdatedIriGenerator;
```
Comment (LucWollants) [inline src/Contributor/ContributorsUpdatedFactory.php:14]: The name can be tweaked, it is not the IRI generator of the update endpoint but of the get endpoint

**Review by LucWollants — APPROVED** [review]

**Review by bertramakers — COMMENTED** [review]
IMHO it would also be okay to have a single ContributorsUpdated message with just the IRI, that requires a lot less code and will also achieve the goal of making it possible for SAPI3 to index the contributors. If we ever need this event for something else where we cannot use the IRI but need the id/type we can easily add those as well (but still keep it 1 class). This is just a suggestion though, the approach with 3 events also works.

To actually get the messages published on AMQP you will need to map the class name(s) of the event(s) here to a content-type: https://github.com/cultuurnet/udb3-backend/blob/7edebd7770f89983a4579a690afcbc1432edddc6/app/AMQP/AMQPPublisherServiceProvider.php#L47-L51 (You can follow the same structure as the existing content-types to invent new ones). The event(s) also need(s) to implement Serializable, but that is already the case I see 👍

**Comment by LucWollants** [comment]
> single ContributorsUpdated message

I thought we had to make the distinction between event, place and organizer to make sure the correct index gets updated?

A functional test is still needed.

**Comment by bertramakers** [comment]
> > single ContributorsUpdated message
> 
> @bertramakers I thought we had to make the distinction between event, place and organizer to make sure the correct index gets updated?
> 
> A functional test is still needed.

Right, overlooked that. That is easier with 3 messages. It's also possible based on the JSON but less "clean" and a bit hard with the current code in SAPI3 I think.

**Comment by JonasVHG** [comment]
> IMHO it would also be okay to have a single ContributorsUpdated message with just the IRI, that requires a lot less code and will also achieve the goal of making it possible for SAPI3 to index the contributors. If we ever need this event for something else where we cannot use the IRI but need the id/type we can easily add those as well (but still keep it 1 class). This is just a suggestion though, the approach with 3 events also works.
> 
> To actually get the messages published on AMQP you will need to map the class name(s) of the event(s) here to a content-type:
> 
> https://github.com/cultuurnet/udb3-backend/blob/7edebd7770f89983a4579a690afcbc1432edddc6/app/AMQP/AMQPPublisherServiceProvider.php#L47-L51
> 
> (You can follow the same structure as the existing content-types to invent new ones). The event(s) also need(s) to implement Serializable, but that is already the case I see 👍

I've tested this with CloudAmqp And I see messages when I locally update contributors.

**Review by LucWollants — APPROVED** [review]

**Pull request merged**